graphs: make init_short_digraph always sort neighbors but without the extra log complexity

cyrilbouvier opened the pull request:
<!-- ^ Please provide a concise and informative title. -->
<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->

This PR improve the `init_short_digraph` function that is used to initialize `StaticSparseCGraph` (used for immutable `Graph` and `DiGraph`).

Before, a boolean parameter `sort_neighbors` was used to specify if we wanted to sort the neighbors or not. It implied an extra `log` in the complexity (as `qsort` was called).
With this PR, the neighbors are always sorted at no extra cost. It is done by appending to the neighbors list the vertices in the correct order so the call to `qsort` is not needed anymore.

This PR partly reverts #38198 and mostly reverts #37662

### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on. For example, -->
<!-- - #12345: short description why this is a dependency -->
<!-- - #34567: ... -->

cyrilbouvier committed:
graphs: init_short_digraph always sort neighbors without extra log complexity

cyrilbouvier committed:
Remove sorting in StaticSparseBackend

cyrilbouvier committed:
Fix some tests whose output depend on the inner representation of static graphs

cyrilbouvier committed:
Fix more tests whose output depend on the inner representation of static graphs

dcoudert commented on src/sage/algebras/lie_algebras/bgg_dual_module.py:
```diff
@@ -202,15 +202,12 @@ def degree_on_basis(self, m):
-             + 2*f[-alpha[5]]*v[Lambda[1] + Lambda[4] - 1/3*Lambda[5]]^*
-             + 3*f[-alpha[4]]*v[Lambda[1] + Lambda[4] - 1/3*Lambda[5]]^*
-             + v[Lambda[1] + Lambda[4] - 1/3*Lambda[5]]^*
+            f[-alpha[2]]^2*f[-alpha[5]]^2*f[-alpha[3]]^3*v[Lambda[1] + Lambda[4] - 1/3*Lambda[5]]^* + 2*f[-alpha[2]]*v[Lambda[1] + Lambda[4] - 1/3*Lambda[5]]^* + 3*f[-alpha[5]]*v[Lambda[1] + Lambda[4] - 1/3*Lambda[5]]^* + v[Lambda[1] + Lambda[4] - 1/3*Lambda[5]]^*
```
can you put this output on multiple lines, similarly to what it was before.

cyrilbouvier replied:
done in e22d5f1f335d23fcdc5e6d7df40d35750339634e

dcoudert commented on src/sage/categories/regular_crystals.py:
```diff
@@ -842,13 +842,8 @@ def dual_equivalence_class(self, index_set=None):
-                 ([[1, 2, 4], [3, 5]], [[1, 2, 3], [4, 5]], 3),
-                 ([[1, 2, 4], [3, 5]], [[1, 2, 3], [4, 5]], 4)]
+                sage: G.edges(sort=True,sort_vertices=True)
+                [([[1, 3, 4], [2, 5]], [[1, 3, 5], [2, 4]], 4), ([[1, 3, 4], [2, 5]], [[1, 2, 4], [3, 5]], 2), ([[1, 2, 5], [3, 4]], [[1, 3, 5], [2, 4]], 2), ([[1, 2, 5], [3, 4]], [[1, 3, 5], [2, 4]], 3), ([[1, 2, 3], [4, 5]], [[1, 2, 4], [3, 5]], 3), ([[1, 2, 3], [4, 5]], [[1, 2, 4], [3, 5]], 4)]
```
please put this output on multiple lines

cyrilbouvier replied:
done in e22d5f1f335d23fcdc5e6d7df40d35750339634e

dcoudert commented on src/sage/graphs/base/static_sparse_graph.pyx:
```diff
@@ -223,119 +226,119 @@ cdef int init_short_digraph(short_digraph g, G, edge_labelled=False,
+            #dereference(g.neighbors[j]) = i
+            if edge_labelled:
+                edge_labels[g.neighbors[j] - g.edges] = e[2]
+            postincrement(g.neighbors[j]) # increment pointer to next item
```
2 spaces before #

cyrilbouvier replied:
done in e22d5f1

dcoudert commented on src/sage/schemes/elliptic_curves/ell_field.py:
```diff
@@ -2305,23 +2305,24 @@ class of curves.  If the j-invariant is not unique in the isogeny
-              'y^2 + x*y + y = x^3 + 4*x + (-6)', None)]
+             ('y^2 + x*y + y = x^3 + 4*x + (-6)',
+              'y^2 + x*y + y = x^3 + (-36)*x + (-70)', None)]
+
```
remove this empty line to avoid warnings about missing `# needs sage.graphs ....`

cyrilbouvier replied:
done in e22d5f1f335d23fcdc5e6d7df40d35750339634e

dcoudert reviewed (changes requested):
This is a smart improvement. Thanks. I have only minor comments.

The tests in `connectivity.pyx` should be fixed since #38535 as been merged.

I agree with you that doctests should not depend on the internal representation of objects. We reduced the number of such doctests, but we still have some.

cyrilbouvier committed:
Improve doctests syntax + respect linter + remove empty line

cyrilbouvier committed:
Merge branch 'develop' into linear-sorted-init-short-digraph

cyrilbouvier committed:
Merge branch 'develop' into linear-sorted-init-short-digraph

dcoudert reviewed (approved):
This is a very nice improvement. Thank you.

LGTM.

cyrilbouvier committed:
Fix doctest by adding a sort

cyrilbouvier committed:
Maybe sorting will fix this doctests...

cyrilbouvier committed:
doctests in bgg_dual_module.py should passed now

I disabled the two tests that depend on `an_element` method on 32-bit:
one should not write tests depending on "random" element.

cyrilbouvier committed:
doctests: more sort and disabling
